tool/grind.dart

```diff
@@ -60,74 +55,67 @@ Directory createTempSync(String prefix) =>
-    await cleanFlutterDir.parent.create(recursive: true);
-    assert(_lockFuture == null);
-    _lockFuture = File(path.join(cleanFlutterDir.parent.path, 'lock'))
-        .openSync(mode: FileMode.write)
-        .lock();
-    await _lockFuture;
-    var lastSynced = File(path.join(cleanFlutterDir.parent.path, 'lastSynced'));
-    var newRepo = FlutterRepo.fromPath(cleanFlutterDir.path, {}, 'clean');
-
-    // We have a repository, but is it up to date?
-    DateTime lastSyncedTime;
-    if (lastSynced.existsSync()) {
-      lastSyncedTime = DateTime.fromMillisecondsSinceEpoch(
-          int.parse(lastSynced.readAsStringSync()));
-    }
-    if (lastSyncedTime == null ||
-        DateTime.now().difference(lastSyncedTime) > Duration(hours: 24)) {
-      // Rebuild the repository.
-      if (cleanFlutterDir.existsSync()) {
-        cleanFlutterDir.deleteSync(recursive: true);
-      }
-      cleanFlutterDir.createSync(recursive: true);
-      await newRepo._init();
-      await lastSynced
-          .writeAsString((DateTime.now()).millisecondsSinceEpoch.toString());
+  var repoCompleter = _cleanFlutterRepo;
+  if (repoCompleter != null) {
+    return repoCompleter.future;
+  }
+
+  // No await is allowed between check of _cleanFlutterRepo and its assignment,
+  // to prevent reentering this function.
+  repoCompleter = Completer();
+
+  // Figure out where the repository is supposed to be and lock updates for
+  // it.
+  await cleanFlutterDir.parent.create(recursive: true);
+  assert(_lockFuture == null);
+  _lockFuture = File(path.join(cleanFlutterDir.parent.path, 'lock'))
+      .openSync(mode: FileMode.write)
+      .lock();
+  await _lockFuture;
+  var lastSynced = File(path.join(cleanFlutterDir.parent.path, 'lastSynced'));
+  var newRepo = FlutterRepo.fromPath(cleanFlutterDir.path, {}, 'clean');
+
+  // We have a repository, but is it up to date?
+  DateTime? lastSyncedTime;
+  if (lastSynced.existsSync()) {
+    lastSyncedTime = DateTime.fromMillisecondsSinceEpoch(
+        int.parse(lastSynced.readAsStringSync()));
+  }
+  if (lastSyncedTime == null ||
+      DateTime.now().difference(lastSyncedTime) > Duration(hours: 24)) {
+    // Rebuild the repository.
+    if (cleanFlutterDir.existsSync()) {
+      cleanFlutterDir.deleteSync(recursive: true);
     }
-    _cleanFlutterRepo.complete(newRepo);
+    cleanFlutterDir.createSync(recursive: true);
+    await newRepo._init();
+    await lastSynced
+        .writeAsString((DateTime.now()).millisecondsSinceEpoch.toString());
   }
-  return _cleanFlutterRepo.future;
+  repoCompleter.complete(newRepo);
+  _cleanFlutterRepo = repoCompleter;
+  return repoCompleter.future;
 }
 
-Directory _dartdocDocsDir;
-
-Directory get dartdocDocsDir => _dartdocDocsDir ??= createTempSync('dartdoc');
-
-Directory _dartdocDocsDirRemote;
-
-Directory get dartdocDocsDirRemote =>
-    _dartdocDocsDirRemote ??= createTempSync('dartdoc_remote');
+late final String _dartdocDocsPath = createTempSync('dartdoc').path;
```

Comment:
Since these are top-level final variables, they are already evaluated lazily. You can remove the `late` here and in similar locations.